fix: Use `IntegerConstant` for loop boundaries in `unrolling`

aakoshh opened the pull request:
# Description

## Problem\*

Resolves #8009 

## Summary\*

Fixes the `get_const_lower_bound` and `get_const_upper_bound` methods in the `unrolling` module to use a new `IntegerConstant` type, which uses the `binary::try_convert_field_element_to_signed_integer` function for `Signed` types to convert them to `i128`, which can correctly represent negative values for our use case. `Unsigned` types are converted to `u128` instead.


## Additional Context

The program in the bug ticket was misleading: `-1 % 5` is 4 according to Python, but -1 according to Rust; it was turned into `4294967295` by the `as u32` operation, not the modulo. I added the cast because the frontend didn't let me use negative literals in the loop.

The negative result for modulo would not necessarily be a problem for the AST story: if we loop from 3..-1, it's not going to do anything, but that's okay, other times it will. Generating a block to declare a value, and then conditionally add the modulo to it to make it positive seems like extreme overkill (it would be unreadable in AST prints). 

The original problem where it turned into 18446744073709551615 did not involve any casting, only a literal in the AST. This literal cannot be used in the for loop in Noir, nor does it allow negatives in the loop, so I added a test directly creating SSA from a hand crafted AST to replicate this. It turned out this is expected behaviour; this number if `u64::MAX`, which is how negative values of `i64` are represented as a `Field` in the SSA. Special care needs to be taken to convert the value back to signed integers. (EDIT: It turns out it can be used if the bounds are declared separately as variables.)

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[For Experimental Features]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [ ] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

TomAFrench commented:
fyi, we currently allow using u128s in loop bounds which may cause issues.

aakoshh commented:
I may have crossed my eyes, but I thought I restricted the index to be 64 bits because Noir wouldn’t let me use 128 bits:

https://github.com/noir-lang/noir/blob/b5918474ccf88fba83d29aa909a8d3048e2d0e71/tooling/ast_fuzzer/src/program/func.rs#L788

vezenovm commented on compiler/noirc_evaluator/src/ssa/opt/unrolling.rs:
```diff
@@ -282,6 +286,17 @@ impl Loop {
+    fn get_const_as_i128(function: &Function, id: ValueId) -> Option<i128> {
+        let (value, typ) = function.dfg.get_numeric_constant_with_type(id)?;
+        if matches!(typ, NumericType::NativeField) {
+            // Shouldn't happen, as fields don't have meaningful ordering.
```
This comment in isolation I think adds confusion as the title and function logic alone indicates we are just attempting to fetch an i128. If we are assuming this function will only ever be called on the jump value of the loop let's comment that on the function itself and replace this with an assertion.

aakoshh replied:
Removed this in favour of an `IntegerConstant` type, which has a constructor returning `None` if it was called with a `NativeField`, so there is no confusion.

vezenovm commented on compiler/noirc_evaluator/src/ssa/opt/unrolling.rs:
```diff
@@ -680,16 +691,14 @@ impl Loop {
 
         Some(BoilerplateStats {
-            iterations: (upper - lower) as usize,
+            iterations: (upper - lower).max(0) as usize,
```
nit: maybe just a brief comment that we do not allow reverse ranges and this will need to be updated in that case

aakoshh replied:
Done 👍

vezenovm commented on compiler/noirc_evaluator/src/ssa/opt/unrolling.rs:
```diff
@@ -1199,6 +1207,27 @@ mod tests {
+        // Looping -4..-1, which should be 3 iterations.
+        let ssa = brillig_unroll_test_case_6470_with_params(
+            "i64",
+            "18446744073709551612",
```
I see it is `-4` from the comment above, but perhaps we could do this through code? Or at least just an additional comment that this is how Noir represents negative numbers.

aakoshh replied:
Changed it to `u64::MAX - 3` and also added a comment.

vezenovm reviewed (commented):
LGTM just some minor nits

vezenovm commented:
> Right. Well currently these two are just optimisations, so hopefully they would just be skipped as None

It does look like as this just used during optimizations we should just return None for u128 range values over i128. So this shouldn't cause any issues, but let's make an issue to follow-up on this (although it will be low priority). Could even have a small unit test that shows an optimization that works will a u32 or u64 bounds but not a certain u128 bound (e.g. like `hoist_array_gets_using_induction_variable_with_const_bound` or a different test that is using the induction variable bounds).

aakoshh commented:
> It does look like as this just used during optimizations we should just return None for u128 range values over i128. 

@asterite pointed out in #8011 that using `u128` range bounds will crash the compiler, which it did. I got confused earlier because of this:

```
error: The value `170141183460469231731687303715884105715` cannot fit into `u32` which has range `0..=4294967295`
   ┌─ src/main.nr:22:14
   │
22 │     for _ in 170141183460469231731687303715884105715..170141183460469231731687303715884105730 {
   │              ---------------------------------------
   │
   = Call stack:
     1. src/main.nr:22:14

Aborting due to 1 previous error
```
I forgot that you can still assign to a `u128` and it will accept it. 

I introduced a new `IntegerConstant` type which can be either `Signed` or `Unsigned` to handle all permissible cases, and updated the unrolling and hoisting to use this type instead of `FieldElement` for handling bounds.

vezenovm commented on compiler/noirc_evaluator/src/ssa/ir/integer.rs:
```diff
@@ -0,0 +1,122 @@
+        match (self, other) {
+            (Self::Signed { value: a, .. }, Self::Signed { value: b, .. }) => a.partial_cmp(b),
+            (Self::Signed { value: a, .. }, Self::Unsigned { value: b, .. }) => {
+                if a.is_negative() || *b > i128::MAX as u128 {
+                    Some(Ordering::Less)
+                } else {
+                    let a: u128 = (*a).try_into().ok()?;
+                    a.partial_cmp(b)
+                }
```
```suggestion
                if a.is_negative() {
                    Some(Ordering::Less)
                } else {
                    (*a).try_into().ok().and_then(|a: u128| a.partial_cmp(b))                
                }
```
nit: we should be able to simplify this if statement

vezenovm commented on compiler/noirc_evaluator/src/ssa/ir/integer.rs:
```diff
@@ -0,0 +1,122 @@
+                }
+            }
+            (Self::Unsigned { value: a, .. }, Self::Signed { value: b, .. }) => {
+                if b.is_negative() || *a > i128::MAX as u128 {
```
```suggestion
                if b.is_negative() {
```
nit: same thing here

vezenovm commented on compiler/noirc_evaluator/src/ssa/ir/integer.rs:
```diff
@@ -0,0 +1,122 @@
+            }
+            (Self::Unsigned { value: a, .. }, Self::Signed { value: b, .. }) => {
+                if b.is_negative() || *a > i128::MAX as u128 {
+                    Some(Ordering::Less)
```
```suggestion
                    Some(Ordering::Greater)
```
`Less` is incorrect no? If `b` is negative the unsigned `a` is always greater.

aakoshh replied:
Ah yes, thanks, copy paste error 🤦

vezenovm commented on compiler/noirc_evaluator/src/ssa/ir/integer.rs:
```diff
@@ -0,0 +1,122 @@
+                if b.is_negative() || *a > i128::MAX as u128 {
+                    Some(Ordering::Less)
+                } else {
+                    let b: u128 = (*b).try_into().ok()?;
+                    a.partial_cmp(&b)
```
```suggestion
                    (*b).try_into().ok().and_then(|b: u128| a.partial_cmp(&b))
```
nit: same thing here

vezenovm reviewed (commented):
Looks great! I like the new type. I just have some minor nits and one correction regarding comparisons in the new type

vezenovm commented:
> `rollup-block-root`	`152` s	`119` s	`1.28`

Pretty surprised to see this large of a regression. We do have large loops that need to be unrolled here. Perhaps we are doing a lot of unnecessary conversions?
